clippy_lints/src/methods/unused_enumerate_index.rs

```diff
@@ -0,0 +1,113 @@
+    let recv_ty = cx.typeck_results().expr_ty(recv);
+    if let Some(recv_ty_defid) = recv_ty.ty_adt_def().map(AdtDef::did)
+        // If we call a method on a `std::iter::Enumerate` instance
+        && match_def_path(cx, recv_ty_defid, &CORE_ITER_ENUMERATE_STRUCT)
+        // If we are calling a method of the `Iterator` trait
+        && is_trait_method(cx, call_expr, sym::Iterator)
+        // And the map argument is a closure
+        && let ExprKind::Closure(closure) = closure_arg.kind
+        && let closure_body = cx.tcx.hir().body(closure.body)
+        // And that closure has one argument ...
+        && let [closure_param] = closure_body.params
+        // .. which is a tuple of 2 elements
+        && let PatKind::Tuple([index, elem], ..) = closure_param.pat.kind
+        // And that the first element (the index) is either `_` or unused in the body
+        && pat_is_wild(cx, &index.kind, closure_body)
+    {
+        // First, try to trigger the lint with our receiver. This will work when chaining methods:
+        // ```
+        // iter.enumerate().map(|(_, x)| x)
+        // ^^^^^^^^^^^^^^^^ `recv`, a call to `std::iter::Iterator::enumerate`
+        // ````
+        if !trigger_lint(cx, recv, closure_param.span, elem.span) {
+            // Otherwise, it may be that we have a binding. Find its initializer:
+            // ```
+            // let binding = iter.enumerate();
+            //               ^^^^^^^^^^^^^^^^ `recv_init_expr`
+            // binding.map(|(_, x)| x)
+            // ^^^^^^^ `recv`, not a call to `std::iter::Iterator::enumerate`
+            let recv_init_expr = expr_or_init(cx, recv);
+            trigger_lint(cx, recv_init_expr, closure_param.span, elem.span);
+
+            // Otherwise, the `Enumerate` probably comes from something that we cannot control.
+            // This would for instance happen with:
+            // ```
+            // external_lib::some_function_returning_enumerate().map(|(_, x)| x)
+            // ```
+        }
+    }
+}
+
+/// Trigger the lint with the suggestion.
+fn trigger_lint(cx: &LateContext<'_>, recv: &Expr<'_>, closure_param_span: Span, elem_span: Span) -> bool {
+    // Check if our `recv` is a call to `std::iter::Iterator::enumerate`.
+    if let ExprKind::MethodCall(_, enumerate_recv, _, enumerate_span) = recv.kind
+        && let Some(enumerate_defid) = cx.typeck_results().type_dependent_def_id(recv.hir_id)
+        && match_def_path(cx, enumerate_defid, &CORE_ITER_ENUMERATE_METHOD)
+    {
+        // If it is, we can suggest removing the tuple from the closure and the preceding call to
+        // `enumerate`, whose span we can get from the `MethodCall`.
+        span_lint_and_then(
+            cx,
+            UNUSED_ENUMERATE_INDEX,
+            enumerate_span,
+            "you seem to use `.enumerate()` and immediately discard the index",
+            |diag| {
+                multispan_sugg(
```

Comment:
FYI (or maybe this was intentional), `multispan_sugg` uses a non-machine-applicable suggestion, meaning that tools like `rustfix` won't try to autofix it (unlike `multispan_sugg_with_applicability(Applicability::MachineApplicable)`).

But there are also subtle edge cases where the current suggestion breaks code. E.g. it currently removes the parameter's type annotation, which can break type inference (#9122 for one example where this has happened before).

We can leave it as is though, it's correct for this to not be machine applicable as is, and it doesn't need to have one. Just thought I'd bring it up in case you intended this to be an autofixable suggestion.

Reply:
This was not at all intentional. I would like it to be machine applicable, if that's at all possible. Is there some kind of testing we could do to find these edge cases?

Reply:
Hm, thinking a bit more about it, the only edge case I can think of is that this removes the parameter's type annotation.
So if we can also replace the tuple type annotation (if it exists) with the second tuple element type then that seems good enough to be machine applicable. I.e.
```rs
.enumerate().map(|(_, v): (usize, i32)| ...);
// to
.enumerate().map(|v: i32| ...);
```
You can then use [`multispan_sugg_with_applicability`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/diagnostics/fn.multispan_sugg_with_applicability.html) instead of `multispan_sugg`, with `Applicability::MachineApplicable`.

Other than just knowing about edge cases from experience I'm not aware of any way we could find them ^^. There is the `lintcheck` tool which runs clippy on a set of crates, and it has an option to also try auto applying suggestions, but I also did a lintcheck run with this PR and it didn't find anything.

Reply:
This should be fixed with my latest push.

The documentation of `multispan_sugg_with_applicability` states:
> rustfix currently doesn’t support the automatic application of suggestions with multiple spans. This is tracked in issue rustfix#141. Suggestions with multiple spans will be silently ignored.

Is this out of date? The issue that is linked is from a repository that was archived, and the issue was closed.

Reply:
Yeah, that looks outdated